crates/telio-nurse/src/aggregator.rs

```diff
@@ -0,0 +1,689 @@
+    // Create a new DataConnectivityAggregator instance
+    #[allow(dead_code)]
+    pub fn new(nurse_features: &FeatureNurse, wg_interface: W) -> ConnectivityDataAggregator<W> {
+        ConnectivityDataAggregator {
+            current_events: parking_lot::RwLock::new(HashMap::new()),
+            unacknowledged_segments: parking_lot::RwLock::new(HashSet::new()),
+            aggregate_relay_events: nurse_features.enable_relay_conn_data,
+            aggregate_nat_traversal_events: nurse_features.enable_nat_traversal_conn_data,
+            wg_interface,
+        }
+    }
+
+    #[allow(dead_code)]
+    pub fn change_peer_state_direct(
+        &mut self,
+        event: AnalyticsEvent,
+        initiator_ep: EndpointProvider,
+        responder_ep: EndpointProvider,
+    ) {
+        self.change_peer_state_common(event, initiator_ep.into(), responder_ep.into())
+    }
+
+    #[allow(dead_code)]
+    pub fn change_peer_state_relayed(&mut self, event: AnalyticsEvent) {
+        self.change_peer_state_common(event, EndpointType::Relay, EndpointType::Relay)
+    }
+
+    fn change_peer_state_common(
+        &mut self,
+        event: AnalyticsEvent,
+        initiator_ep: EndpointType,
+        responder_ep: EndpointType,
+    ) {
+        if !self.aggregate_nat_traversal_events {
+            return;
+        }
+
+        match self.current_events.write().entry(event.public_key) {
+            std::collections::hash_map::Entry::Occupied(mut entry) => {
+                let (old_event, old_state) = entry.get();
+
+                let connection_data = match old_state {
+                    ConnectionState::Relay(_) => {
+                        telio_log_warn!("Peer event for the relay server pubkey - discarding");
+                        return;
+                    }
+                    ConnectionState::Peer(ep1, ep2) => {
+                        if old_event.peer_state == event.peer_state
+                            && *ep1 == initiator_ep
+                            && *ep2 == responder_ep
+                        {
+                            // Peers state remains unchanged - skipping the segment addition
+                            return;
+                        }
+                        ConnectionData::Peer(PeerConnectionData {
+                            initiator_ep_type: *ep1,
+                            reciever_ep_type: *ep2,
+                            rx_bytes: event.rx_bytes - old_event.rx_bytes,
+                            tx_bytes: event.tx_bytes - old_event.tx_bytes,
+                        })
+                    }
+                };
+
+                // Just create a new segment here
+                self.unacknowledged_segments
+                    .write()
+                    .insert(ConnectionSegmentData {
+                        node: event.public_key,
+                        start: old_event.timestamp.into(),
+                        length: event.timestamp - old_event.timestamp,
```

Comment:
:thinking: This would produce invalid length in case of device sleeping, as we would simple stretch the state during sleep, while in reality, we should consider it as being offline. But it is rather hard to make it "correct". In order to avoid next task blocking, this sleeping behavior may be solved in different task/PR.

Reply:
After a little discussion, we have decided to offload handling of sleeping behavior into separate task